### PR TITLE
Onboarding flow

### PR DESCRIPTION
**Pull request by alanctkc**
@dopey @maraino @sourishkrout This is the minimal wiring necessary to connect `step-ca` to the new onboarding/bootstrapping flow.

It should be enough that someone could pick it up, refactor it, add in the actual `init` and listening behavior, etc.

It hits the api to retrieve the configuration settings filled into the onboarding forms, then it posts back with the fingerprint of the CA's root cert, which moves the onboarding guide to the next step and fills the fingerprint into a `step ca bootstrap` command further into the flow.

@sourishkrout Feel free to pull it down and give it a whirl against the onboarding flow in local. In the actual released version, it will point to `api.smallstep.com` instead of local.

Example usage:

```console
$ step-ca start _AhQZCMXHVBKrSC6AAAC
Connecting to onboarding guide...

Connected! Initializing step-ca with the following configuration...

Name: Smallstep
DNS: localhost
Address: :8443
Provisioner Password: abcdef1234567890

Initialized!
Step CA has been started. Please return to the onboarding guide in your browser to continue.
```

I also envision `step-ca start` eventually being the canonical way to start the CA, with an API something like this:

```console
$ step-ca start # start with default config at ~/.step/config/ca.json
$ step-ca start --config ~/path/to/ca.json # non-default path
$ step-ca start <token> # configure from onboarding flow and start
```

Thoughts?

### CODE REVIEW

**Comment by sourishkrout** [comment]
Awesome, @alanctkc! Let's review the code and make plans how to tackle this at tomorrow's meeting!

File: cmd/step-ca/main.go [hunk 1]
```diff
@@ -179,6 +189,67 @@ intermediate private key.`,
...
+				if err != nil {
+					log.Fatal(err)
+				}
+
+				body, err := ioutil.ReadAll(res.Body)
+				if err != nil {
+					log.Fatal(err)
+				}
+
+				configuration := config{}
```
Comment (maraino) [inline cmd/step-ca/main.go]: I would rename the `config` struct as configuration and the variable as `config`, looks better in my opinion.
```go
var config configuration // or config := configuration{}
```

File: cmd/step-ca/main.go [hunk 2]
```diff
@@ -179,6 +189,67 @@ intermediate private key.`,
...
+			// TODO this should accept an optional config parameter that defaults to ~/.step/config/ca.json
+			// as well as an optional token parameter for connecting to the onboarding flow
+			Action: func(c *cli.Context) error {
+				fmt.Printf("Connecting to onboarding guide...\n\n")
+
+				token := c.Args().Get(0)
+
+				res, err := http.Get("http://localhost:3002/onboarding/" + token)
+				if err != nil {
+					log.Fatal(err)
```
[the remark this answers is not in the document]
Comment (maraino) [reply cmd/step-ca/main.go]: Do the same on all errors.

File: cmd/step-ca/main.go [hunk 3]
```diff
@@ -179,6 +189,67 @@ intermediate private key.`,
...
+				}
+
+				fmt.Printf("Connected! Initializing step-ca with the following configuration...\n\n")
+				fmt.Printf("Name: %s\n", configuration.Name)
+				fmt.Printf("DNS: %s\n", configuration.DNS)
+				fmt.Printf("Address: %s\n", configuration.Address)
+				// TODO generate this password
+				fmt.Printf("Provisioner Password: abcdef1234567890\n\n")
```
Comment (maraino) [inline cmd/step-ca/main.go]: You can use `func Alphanumeric(length int) string` from `github.com/smallstep/cli/crypto/randutil`

File: cmd/step-ca/main.go [hunk 4]
```diff
@@ -179,6 +189,67 @@ intermediate private key.`,
...
+				req.Header.Set("Content-Type", "application/json")
+				if err != nil {
+					log.Fatal(err)
+				}
+
+				client := &http.Client{}
+				resp, err := client.Do(req)
+				if err != nil {
+					log.Fatal(err)
+				}
```
Comment (maraino) [inline cmd/step-ca/main.go]: Use `func Post(url, contentType string, body io.Reader) (resp *Response, err error)`, is simpler.

File: cmd/step-ca/main.go [hunk 5]
```diff
@@ -179,6 +189,67 @@ intermediate private key.`,
 				return nil
 			},
 		},
+		{
+			Name:  "start",
+			Usage: "Starts step-ca with the (optional) specified configuration",
+			// TODO this should accept an optional config parameter that defaults to ~/.step/config/ca.json
+			// as well as an optional token parameter for connecting to the onboarding flow
+			Action: func(c *cli.Context) error {
```
[the remark this answers is not in the document]
Comment (maraino) [reply cmd/step-ca/main.go]: I've just noticed we already have one, just use a different name if you want to keep the other one.

File: cmd/step-ca/main.go [hunk 6]
```diff
@@ -179,6 +189,67 @@ intermediate private key.`,
...
+				}
+
+				client := &http.Client{}
+				resp, err := client.Do(req)
+				if err != nil {
+					log.Fatal(err)
+				}
+				resp.Body.Close()
+
+				fmt.Printf("Initialized!\n")
```
Comment (maraino) [inline cmd/step-ca/main.go]: You can use `fmt.Println("Initialized!")` to avoid the escaped characters.

File: cmd/step-ca/main.go [hunk 7]
```diff
@@ -179,6 +189,67 @@ intermediate private key.`,
...
+				fmt.Printf("Step CA has been started. Please return to the onboarding guide in your browser to continue.\n")
+				for {
+					time.Sleep(1 * time.Second)
+				}
```
Comment (maraino) [inline cmd/step-ca/main.go]: I suppose this is for testing purposes. See current `startAction` function to see how to start the server.

**Comment by alanctkc** [comment]
Let's change the command to `step-ca onboard <token>` for now. :+1:

File: cmd/step-ca/main.go [hunk 8]
```diff
@@ -179,6 +189,67 @@ intermediate private key.`,
...
+			Name:  "start",
+			Usage: "Starts step-ca with the (optional) specified configuration",
+			// TODO this should accept an optional config parameter that defaults to ~/.step/config/ca.json
+			// as well as an optional token parameter for connecting to the onboarding flow
+			Action: func(c *cli.Context) error {
+				fmt.Printf("Connecting to onboarding guide...\n\n")
+
+				token := c.Args().Get(0)
+
+				res, err := http.Get("http://localhost:3002/onboarding/" + token)
```
Comment (maraino) [inline cmd/step-ca/main.go]: I think we will need an environment variable for the URL.

**Comment by alanctkc** [comment]
This is looking great to me. Maybe I'll tag @dopey for a final review if ready?

File: commands/app.go [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+	"github.com/smallstep/certificates/ca"
+	"github.com/smallstep/cli/errs"
+	"github.com/urfave/cli"
+)
+
+// AppCommand is the action used as the top action.
+var AppCommand = cli.Command{
+	Name:   "start",
+	Action: appAction,
+	Flags: []cli.Flag{
```
Comment (dopey) [inline commands/app.go]: usage text? description?

Comment (maraino) [reply commands/app.go]: This is a virtual command for the cli.App, I basically moved the code to a new file. The description and all that is in the main.go

File: commands/onboard.go [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+
+func init() {
+	command.Register(cli.Command{
+		Name:      "onboard",
+		Usage:     "configure and run step-ca from the onboarding guide",
+		UsageText: "**step-ca onboard** <token>",
+		Action:    onboardAction,
+		Description: `**step-ca onboard** configures step certificates using the onboarding guide.
+
+Open https://smallstep.com/onboarding in your browser and start the CA with the
```
Comment (dopey) [inline commands/onboard.go]: is it api.smallstep.com or smallstep.com? earlier you used api.smallstep.com

Comment (maraino) [reply commands/onboard.go]: https://smallstep.com/onboarding is the web app, api.smallstep.com is the endpoint step-ca uses to get the configuration.

**Pull request merged**